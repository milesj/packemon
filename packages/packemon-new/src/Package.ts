import { Path, toArray } from '@boost/common';
import { optimal } from '@boost/common/optimal';
import { createDebugger, Debugger } from '@boost/debug';
import {
	DEFAULT_FORMATS,
	EXCLUDE,
	EXTENSIONS,
	FORMATS_BROWSER,
	FORMATS_NATIVE,
	FORMATS_NODE,
	NODE_SUPPORTED_VERSIONS,
	NPM_SUPPORTED_VERSIONS,
	SUPPORT_PRIORITY,
} from './constants';
import { packemonBlueprint } from './schemas';
import { PackageConfig, PackemonPackage, PackemonPackageConfig } from './types';

export class Package {
	// readonly artifacts: Artifact[] = [];

	readonly configs: PackageConfig[] = [];

	readonly debug!: Debugger;

	readonly json: PackemonPackage;

	readonly jsonPath: Path;

	readonly path: Path;

	constructor(path: Path, contents: PackemonPackage) {
		this.path = path;
		this.jsonPath = this.path.append('package.json');
		this.json = contents;
		this.debug = createDebugger(['packemon', 'package', this.getSlug()]);
	}

	async clean(): Promise<void> {
		this.debug('Cleaning build artifacts');

		// await Promise.all(this.artifacts.map((artifact) => artifact.cleanup()));
	}

	getName(): string {
		return this.json.name;
	}

	getSlug(): string {
		return this.path.name(true);
	}

	setConfigs(configs: PackemonPackageConfig[]) {
		configs.forEach((cfg) => {
			const config = optimal(packemonBlueprint, {
				name: this.getName(),
			}).validate(cfg);

			// eslint-disable-next-line complexity
			toArray(config.platform).forEach((platform) => {
				let { api, bundle } = config;
				let formats = config.format ? [config.format] : [];

				switch (platform) {
					case 'native':
						formats = formats.filter((format) => (FORMATS_NATIVE as string[]).includes(format));

						if (formats.length === 0) {
							formats.push(DEFAULT_FORMATS.native);
						}
						break;

					case 'node':
						if (cfg.api === undefined) {
							api = 'public';
						}

						if (cfg.bundle === undefined) {
							bundle = false;
						}

						formats = formats.filter((format) => (FORMATS_NODE as string[]).includes(format));

						if (formats.length === 0) {
							formats.push(DEFAULT_FORMATS.node);
						}
						break;

					default:
						formats = formats.filter((format) => (FORMATS_BROWSER as string[]).includes(format));

						if (formats.length === 0) {
							formats.push(DEFAULT_FORMATS.browser);
						}

						// Auto-support lib builds for test environments
						if (formats.includes('esm') && !formats.includes('lib')) {
							formats.push('lib');
						}

						if (config.namespace && !formats.includes('umd')) {
							formats.push('umd');
						}
						break;
				}

				this.configs.push({
					api,
					bundle,
					externals: toArray(config.externals),
					formats,
					inputs: config.inputs,
					namespace: config.namespace,
					platform,
					support: config.support,
				});
			});
		});
	}
}
