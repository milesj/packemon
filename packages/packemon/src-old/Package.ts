import glob from 'fast-glob';
import fs from 'fs-extra';
import semver from 'semver';
import { deepMerge, isObject, Memoize, PackageStructure, Path, toArray } from '@boost/common';
import { optimal } from '@boost/common/optimal';
import { createDebugger, Debugger } from '@boost/debug';
import { Artifact } from './Artifact';
import { CodeArtifact } from './CodeArtifact';
import {
	DEFAULT_FORMATS,
	EXCLUDE,
	EXTENSIONS,
	FORMATS_BROWSER,
	FORMATS_NATIVE,
	FORMATS_NODE,
	NODE_SUPPORTED_VERSIONS,
	NPM_SUPPORTED_VERSIONS,
	SUPPORT_PRIORITY,
} from './constants';
import { loadModule } from './helpers/loadModule';
import { sortExports } from './helpers/sortExports';
import { Project } from './Project';
import { packemonBlueprint } from './schemas';
import {
	BuildOptions,
	ConfigFile,
	FeatureFlags,
	InputMap,
	PackageConfig,
	PackageExportPaths,
	PackageExports,
	PackemonPackage,
	PackemonPackageConfig,
	TSConfigStructure,
} from './types';
import { TypesArtifact } from './TypesArtifact';

export class Package {
	isComplete(): boolean {
		return this.artifacts.every((artifact) => artifact.isComplete());
	}

	isRunning(): boolean {
		return this.artifacts.some((artifact) => artifact.isRunning());
	}

	protected addEngines() {
		const artifact = (this.artifacts as CodeArtifact[])
			.filter((art) => art instanceof CodeArtifact)
			.filter((art) => art.platform === 'node')
			.reduce<CodeArtifact | null>(
				(oldest, art) =>
					!oldest || SUPPORT_PRIORITY[art.support] < SUPPORT_PRIORITY[oldest.support]
						? art
						: oldest,
				null,
			);

		if (!artifact) {
			return;
		}

		this.debug('Adding `engines` to `package.json`');

		const pkg = this.packageJson;

		if (!pkg.engines) {
			pkg.engines = {};
		}

		Object.assign(pkg.engines, {
			node: `>=${NODE_SUPPORTED_VERSIONS[artifact.support]}`,
			npm: toArray(NPM_SUPPORTED_VERSIONS[artifact.support])
				.map((v) => `>=${v}`)
				.join(' || '),
		});
	}

	protected addEntryPoints() {
		this.debug('Adding entry points to `package.json`');

		let mainEntry: string | undefined;
		let moduleEntry: string | undefined;
		let browserEntry: string | undefined;
		let buildCount = 0;

		// eslint-disable-next-line complexity
		this.artifacts.forEach((artifact) => {
			// Build files
			if (artifact instanceof CodeArtifact) {
				const mainEntryName = artifact.inputs.index ? 'index' : Object.keys(artifact.inputs)[0];

				buildCount += artifact.builds.length;

				// Generate `main`, `module`, and `browser` fields
				if (!mainEntry || (artifact.platform === 'node' && mainEntryName === 'index')) {
					mainEntry = artifact.findEntryPoint(['lib', 'cjs', 'mjs', 'esm'], mainEntryName);
				}

				if (!moduleEntry || (artifact.platform === 'browser' && mainEntryName === 'index')) {
					moduleEntry = artifact.findEntryPoint(['esm'], mainEntryName);
				}

				// Only include when we share a lib with another platform
				if (!browserEntry && artifact.platform === 'browser') {
					browserEntry = artifact.findEntryPoint(
						artifact.sharedLib ? ['lib', 'umd'] : ['umd'],
						mainEntryName,
					);
				}

				// Generate `bin` field
				if (
					artifact.inputs.bin &&
					artifact.platform === 'node' &&
					!isObject(this.packageJson.bin)
				) {
					this.packageJson.bin = artifact.findEntryPoint(['lib', 'cjs', 'mjs'], 'bin');
				}
			}

			// Type declarations
			if (artifact instanceof TypesArtifact) {
				const mainEntryName = artifact.builds.some((build) => build.outputName === 'index')
					? 'index'
					: artifact.builds[0].outputName;

				this.packageJson.types = artifact.findEntryPoint(mainEntryName);
			}
		});

		if (mainEntry) {
			this.packageJson.main = mainEntry;

			// Only set when we have 1 build, otherwise its confusing
			if (buildCount === 1) {
				if (mainEntry.includes('mjs/') || mainEntry.includes('esm/')) {
					this.packageJson.type = 'module';
				} else if (mainEntry.includes('cjs/')) {
					this.packageJson.type = 'commonjs';
				}
			}
		}

		if (moduleEntry) {
			this.packageJson.module = moduleEntry;
		}

		if (browserEntry && !isObject(this.packageJson.browser)) {
			this.packageJson.browser = browserEntry;
		}
	}

	protected addExports() {
		this.debug('Adding `exports` to `package.json`');

		let exportMap: PackageExports = {
			'./package.json': './package.json',
		};

		this.artifacts.forEach((artifact) => {
			Object.entries(artifact.getPackageExports()).forEach(([path, conditions]) => {
				if (!conditions) {
					return;
				}

				if (!exportMap[path]) {
					exportMap[path] = conditions;
					return;
				}

				if (typeof exportMap[path] === 'string') {
					exportMap[path] = { default: exportMap[path] };
				}

				exportMap[path] = deepMerge<PackageExportPaths, PackageExportPaths>(
					exportMap[path] as PackageExportPaths,
					typeof conditions === 'string' ? { default: conditions } : conditions,
				);
			});
		});

		exportMap = sortExports(exportMap);

		if (isObject(this.packageJson.exports)) {
			Object.assign(this.packageJson.exports, exportMap);
		} else {
			this.packageJson.exports = exportMap as PackageStructure['exports'];
		}
	}

}
