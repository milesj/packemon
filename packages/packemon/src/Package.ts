/* eslint-disable require-atomic-updates, no-param-reassign, @typescript-eslint/member-ordering */

import glob from 'fast-glob';
import fs from 'fs-extra';
import { isObject, Memoize, optimal, PackageStructure, Path, toArray } from '@boost/common';
import { createDebugger, Debugger } from '@boost/debug';
import { Artifact } from './Artifact';
import { CodeArtifact } from './CodeArtifact';
import {
	DEFAULT_FORMATS,
	FORMATS_BROWSER,
	FORMATS_NATIVE,
	FORMATS_NODE,
	NODE_SUPPORTED_VERSIONS,
	NPM_SUPPORTED_VERSIONS,
	SUPPORT_PRIORITY,
} from './constants';
import { loadModule } from './helpers/loadModule';
import { Project } from './Project';
import { packemonBlueprint } from './schemas';
import {
	BuildOptions,
	FeatureFlags,
	InputMap,
	PackageConfig,
	PackageExports,
	PackemonPackage,
	PackemonPackageConfig,
	TSConfigStructure,
} from './types';
import { TypesArtifact } from './TypesArtifact';

export class Package {
	readonly artifacts: Artifact[] = [];

	readonly configs: PackageConfig[] = [];

	readonly debug!: Debugger;

	readonly packageJson: PackemonPackage;

	readonly packageJsonPath: Path;

	readonly path: Path;

	readonly project: Project;

	root: boolean = false;

	constructor(project: Project, path: Path, contents: PackemonPackage) {
		this.project = project;
		this.path = path;
		this.packageJsonPath = this.path.append('package.json');
		this.packageJson = contents;
		this.debug = createDebugger(['packemon', 'package', this.getSlug()]);
	}

	addArtifact(artifact: Artifact): Artifact {
		this.artifacts.push(artifact);

		artifact.startup();

		return artifact;
	}

	async build(options: BuildOptions): Promise<void> {
		this.debug('Building artifacts');

		// Build artifacts in parallel
		await Promise.all(
			this.artifacts.map(async (artifact) => {
				const start = Date.now();

				try {
					artifact.state = 'building';

					await artifact.build(options);

					artifact.state = 'passed';
				} catch (error: unknown) {
					artifact.state = 'failed';

					throw error;
				} finally {
					artifact.buildResult.time = Date.now() - start;
				}
			}),
		);

		// Add package entry points based on artifacts
		this.addEntryPoints();

		// Add package `engines` based on artifacts
		if (options.addEngines) {
			this.addEngines();
		}

		// Add package `exports` based on artifacts
		if (options.addExports) {
			this.addExports();
		}

		// Sync `package.json` in case it was modified
		await this.syncPackageJson();
	}

	async cleanup(): Promise<void> {
		this.debug('Cleaning build artifacts');

		await Promise.all(this.artifacts.map((artifact) => artifact.cleanup()));
	}

	getName(): string {
		return this.packageJson.name;
	}

	@Memoize()
	// eslint-disable-next-line complexity
	getFeatureFlags(): FeatureFlags {
		this.debug('Loading feature flags');

		const flags: FeatureFlags =
			this.root || !this.project.isWorkspacesEnabled()
				? {}
				: this.project.rootPackage.getFeatureFlags();

		flags.workspaces = this.project.workspaces;

		// React
		if (this.project.rootPackage.hasDependency('react') || this.hasDependency('react')) {
			flags.react = true;

			this.debug(' - React');
		}

		// TypeScript
		const tsConfig = this.tsconfigJson ?? this.project.rootPackage.tsconfigJson;

		if (
			this.project.rootPackage.hasDependency('typescript') ||
			this.hasDependency('typescript') ||
			tsConfig
		) {
			flags.typescript = true;
			flags.decorators = Boolean(tsConfig?.options.experimentalDecorators);
			flags.strict = Boolean(tsConfig?.options.strict);

			this.debug(
				' - TypeScript (%s, %s)',
				flags.strict ? 'strict' : 'non-strict',
				flags.decorators ? 'decorators' : 'non-decorators',
			);
		}

		// Flow
		const flowconfigPath = this.project.root.append('.flowconfig');

		if (
			this.project.rootPackage.hasDependency('flow-bin') ||
			this.hasDependency('flow-bin') ||
			flowconfigPath.exists()
		) {
			flags.flow = true;

			this.debug(' - Flow');
		}

		return flags;
	}

	getSlug(): string {
		return this.path.name(true);
	}

	@Memoize()
	getSourceFiles(): string[] {
		return glob.sync('src/**/*.{js,jsx,ts,tsx}', {
			absolute: true,
			cwd: this.path.path(),
			onlyFiles: true,
		});
	}

	hasDependency(name: string): boolean {
		const pkg = this.packageJson;

		return Boolean(
			pkg.dependencies?.[name] ??
				pkg.devDependencies?.[name] ??
				pkg.peerDependencies?.[name] ??
				pkg.optionalDependencies?.[name],
		);
	}

	isComplete(): boolean {
		return this.artifacts.every((artifact) => artifact.isComplete());
	}

	isRunning(): boolean {
		return this.artifacts.some((artifact) => artifact.isRunning());
	}

	setConfigs(configs: PackemonPackageConfig[]) {
		configs.forEach((cfg) => {
			const config = optimal(cfg, packemonBlueprint, {
				name: this.getName(),
			});

			toArray(config.platform).forEach((platform) => {
				let { bundle } = config;
				let formats = [...toArray(config.format)];
				const isEmpty = formats.length === 0;

				switch (platform) {
					case 'native':
						if (isEmpty) {
							formats.push(...DEFAULT_FORMATS.native);
						} else {
							formats = formats.filter((format) => (FORMATS_NATIVE as string[]).includes(format));
						}
						break;

					case 'node':
						if (cfg.bundle === undefined) {
							bundle = false;
						}

						if (isEmpty) {
							formats.push(...DEFAULT_FORMATS.node);
						} else {
							formats = formats.filter((format) => (FORMATS_NODE as string[]).includes(format));
						}
						break;

					case 'browser':
					default:
						if (isEmpty) {
							formats.push(...DEFAULT_FORMATS.browser);

							if (config.namespace) {
								formats.push('umd');
							}
						} else {
							formats = formats.filter((format) => (FORMATS_BROWSER as string[]).includes(format));
						}
						break;
				}

				this.configs.push({
					bundle,
					externals: toArray(config.externals),
					formats,
					inputs: config.inputs,
					namespace: config.namespace,
					platform,
					support: config.support,
				});
			});
		});
	}

	async syncPackageJson() {
		await fs.writeJson(this.packageJsonPath.path(), this.packageJson, { spaces: 2 });
	}

	@Memoize()
	get tsconfigJson(): TSConfigStructure | undefined {
		const tsconfigJsonPath = this.path.append('tsconfig.json');

		if (!tsconfigJsonPath.exists()) {
			return undefined;
		}

		const ts = loadModule(
			'typescript',
			'TypeScript is required for config loading.',
		) as typeof import('typescript');

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		const { config, error } = ts.readConfigFile(tsconfigJsonPath.path(), (name) =>
			fs.readFileSync(name, 'utf8'),
		);

		const host = {
			getCanonicalFileName: (fileName: string) => fileName,
			getCurrentDirectory: () => ts.sys.getCurrentDirectory(),
			getNewLine: () => ts.sys.newLine,
		};

		// istanbul ignore next
		if (error) {
			throw new Error(ts.formatDiagnostic(error, host));
		}

		const result = ts.parseJsonConfigFileContent(
			config,
			ts.sys,
			this.path.path(),
			{},
			tsconfigJsonPath.path(),
		);

		// istanbul ignore next
		if (result.errors.length > 0) {
			throw new Error(ts.formatDiagnostics(result.errors, host));
		}

		return result;
	}

	protected addEngines() {
		const artifact = (this.artifacts as CodeArtifact[])
			.filter((art) => art instanceof CodeArtifact)
			.filter((art) => art.platform === 'node')
			.reduce<CodeArtifact | null>(
				(oldest, art) =>
					!oldest || SUPPORT_PRIORITY[art.support] < SUPPORT_PRIORITY[oldest.support]
						? art
						: oldest,
				null,
			);

		if (!artifact) {
			return;
		}

		this.debug('Adding `engines` to `package.json`');

		const pkg = this.packageJson;

		if (!pkg.engines) {
			pkg.engines = {};
		}

		Object.assign(pkg.engines, {
			node: `>=${NODE_SUPPORTED_VERSIONS[artifact.support]}`,
			npm: toArray(NPM_SUPPORTED_VERSIONS[artifact.support])
				.map((v) => `>=${v}`)
				.join(' || '),
		});
	}

	protected addEntryPoints() {
		this.debug('Adding entry points to `package.json`');

		let mainEntry = '';
		let moduleEntry = '';
		let browserEntry = '';
		const files = new Set<string>(this.packageJson.files);

		// eslint-disable-next-line complexity
		this.artifacts.forEach((artifact) => {
			// Build files
			if (artifact instanceof CodeArtifact) {
				// Generate `main`, `module`, and `browser` fields
				if (artifact.inputs.index) {
					if (!mainEntry || artifact.platform === 'node') {
						mainEntry = artifact.findEntryPoint(['lib', 'cjs', 'mjs', 'esm'], 'index');
					}

					if (!moduleEntry) {
						moduleEntry = artifact.findEntryPoint(['esm', 'mjs'], 'index');
					}

					// Only include when we share a lib with another platform
					if (!browserEntry && artifact.platform === 'browser') {
						browserEntry = artifact.findEntryPoint(
							artifact.sharedLib ? ['lib', 'umd'] : ['umd'],
							'index',
						);
					}
				}

				// Generate `bin` field
				if (
					artifact.inputs.bin &&
					artifact.platform === 'node' &&
					!isObject(this.packageJson.bin)
				) {
					this.packageJson.bin = artifact.findEntryPoint(['lib', 'cjs', 'mjs'], 'bin');
				}

				// Generate `files` list
				artifact.builds.forEach(({ format }) => {
					files.add(`${format}/**/*.{${artifact.getBuildOutput(format).ext},map}`);
				});

				files.add(`src/**/*.{${this.getSourceFileExts(artifact.inputs)}}`);
			}

			// Type declarations
			if (artifact instanceof TypesArtifact) {
				this.packageJson.types = artifact.findEntryPoint('index');
				files.add('dts/**/*.d.ts');
			}
		});

		if (mainEntry) {
			this.packageJson.main = mainEntry;

			if (mainEntry.includes('mjs/') || mainEntry.includes('esm/')) {
				this.packageJson.type = 'module';
			} else if (mainEntry.includes('cjs/')) {
				this.packageJson.type = 'commonjs';
			}
		}

		if (moduleEntry) {
			this.packageJson.module = moduleEntry;
		}

		if (browserEntry && !isObject(this.packageJson.browser)) {
			this.packageJson.browser = browserEntry;
		}

		this.packageJson.files = [...files].sort();
	}

	protected addExports() {
		this.debug('Adding `exports` to `package.json`');

		const exportMap: PackageExports = {
			'./package.json': './package.json',
		};

		this.artifacts.forEach((artifact) => {
			Object.entries(artifact.getPackageExports()).forEach(([basePath, conditions]) => {
				const path = basePath.replace('/index', '');

				if (!exportMap[path]) {
					exportMap[path] = {};
				} else if (typeof exportMap[path] === 'string') {
					exportMap[path] = { default: exportMap[path] };
				}

				Object.assign(exportMap[path], conditions);
			});
		});

		if (isObject(this.packageJson.exports)) {
			Object.assign(this.packageJson.exports, exportMap);
		} else {
			this.packageJson.exports = exportMap as PackageStructure['exports'];
		}
	}

	protected getSourceFileExts(inputs: InputMap): string[] {
		const sourceExts = Object.values(inputs).map((inputFile) => new Path(inputFile).ext(true));
		const exts = new Set(sourceExts);

		// Include sibling file extensions
		sourceExts.forEach((sourceExt) => {
			switch (sourceExt) {
				case 'js':
					exts.add('jsx');
					break;

				case 'jsx':
				case 'cjs':
					exts.add('js');
					break;

				case 'ts':
					exts.add('tsx');
					break;

				case 'tsx':
					exts.add('ts');
					break;

				// no default
			}
		});

		const list = [...exts].sort();

		// Always be last
		list.push('json');

		return list;
	}
}
