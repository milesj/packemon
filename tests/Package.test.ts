import fs from 'fs-extra';
import { Path } from '@boost/common';
import { getFixturePath } from '@boost/test-utils';
import Package from '../src/Package';
import Project from '../src/Project';
import { TestArtifact } from './helpers';

describe('Package', () => {
  const fixturePath = getFixturePath('workspaces-feature-flags');
  let project: Project;
  let pkg: Package;

  function createTestArtifacts() {
    const a = new TestArtifact(pkg, []);
    const b = new TestArtifact(pkg, []);
    const c = new TestArtifact(pkg, []);

    return [a, b, c];
  }

  function loadPackage(name: string, customProject?: Project): Package {
    const pkgPath = new Path(fixturePath, 'packages', name);

    return new Package(
      customProject || project,
      pkgPath,
      fs.readJsonSync(pkgPath.append('package.json').path()),
    );
  }

  beforeEach(() => {
    project = new Project(fixturePath);
    pkg = loadPackage('common');
  });

  it('sets properties on instantiation', () => {
    expect(pkg.project).toBe(project);
    expect(pkg.path).toEqual(new Path(fixturePath, 'packages/common'));
    expect(pkg.packageJsonPath).toEqual(new Path(fixturePath, 'packages/common/package.json'));
    expect(pkg.packageJson).toEqual({
      name: 'flag-common',
    });
  });

  describe('addArtifact()', () => {
    it('adds an artifact and calls `startup`', () => {
      const artifact = new TestArtifact(pkg, []);
      const spy = jest.spyOn(artifact, 'startup');

      pkg.addArtifact(artifact);

      expect(pkg.artifacts).toEqual([artifact]);
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('build()', () => {
    let writeSpy: jest.SpyInstance;

    beforeEach(() => {
      writeSpy = jest.spyOn(fs, 'writeJson').mockImplementation();
    });

    afterEach(() => {
      writeSpy.mockRestore();
    });

    it('calls `build` on each artifact', async () => {
      const [a, b, c] = createTestArtifacts();
      const aSpy = jest.spyOn(a, 'build');
      const bSpy = jest.spyOn(b, 'build');
      const cSpy = jest.spyOn(c, 'build');

      pkg.addArtifact(a);
      pkg.addArtifact(b);
      pkg.addArtifact(c);

      await pkg.build({ concurrency: 1 });

      expect(aSpy).toHaveBeenCalledWith({ concurrency: 1 });
      expect(bSpy).toHaveBeenCalledWith({ concurrency: 1 });
      expect(cSpy).toHaveBeenCalledWith({ concurrency: 1 });
    });

    it('calls `preBuild` and `postBuild` when building', async () => {
      const artifact = new TestArtifact(pkg, []);
      const preSpy = jest.spyOn(artifact, 'build');
      const postSpy = jest.spyOn(artifact, 'build');

      pkg.addArtifact(artifact);

      await pkg.build({ addEngines: true });

      expect(preSpy).toHaveBeenCalledWith({ addEngines: true });
      expect(postSpy).toHaveBeenCalledWith({ addEngines: true });
    });

    it('sets passed state and result time', async () => {
      const artifact = new TestArtifact(pkg, []);

      pkg.addArtifact(artifact);

      expect(artifact.state).toBe('pending');
      expect(artifact.buildResult.time).toBe(0);

      await pkg.build({});

      expect(artifact.state).toBe('passed');
      expect(artifact.buildResult.time).not.toBe(0);
    });

    it('sets failed state and result time on error', async () => {
      const artifact = new TestArtifact(pkg, []);

      jest.spyOn(artifact, 'postBuild').mockImplementation(() => {
        throw new Error('Whoops');
      });

      pkg.addArtifact(artifact);

      expect(artifact.state).toBe('pending');
      expect(artifact.buildResult.time).toBe(0);

      try {
        await pkg.build({});
      } catch (error) {
        expect(error).toEqual(new Error('Whoops'));
      }

      expect(artifact.state).toBe('failed');
      expect(artifact.buildResult.time).not.toBe(0);
    });

    it('syncs `package.json` when done building', async () => {
      const artifact = new TestArtifact(pkg, []);
      const spy = jest.spyOn(pkg, 'syncPackageJson');

      pkg.addArtifact(artifact);

      await pkg.build({});

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('cleanup()', () => {
    it('calls `cleanup` on each artifact', async () => {
      const [a, b, c] = createTestArtifacts();
      const aSpy = jest.spyOn(a, 'cleanup');
      const bSpy = jest.spyOn(b, 'cleanup');
      const cSpy = jest.spyOn(c, 'cleanup');

      pkg.addArtifact(a);
      pkg.addArtifact(b);
      pkg.addArtifact(c);

      await pkg.cleanup();

      expect(aSpy).toHaveBeenCalled();
      expect(bSpy).toHaveBeenCalled();
      expect(cSpy).toHaveBeenCalled();
    });
  });

  describe('getName()', () => {
    it('returns `name` from `package.json`', () => {
      expect(pkg.getName()).toBe('flag-common');
    });
  });

  describe('getSlug()', () => {
    it('returns package folder name', () => {
      expect(pkg.getSlug()).toBe('common');
    });
  });

  describe('getFeatureFlags()', () => {
    it('inherits workspaces from root project', () => {
      project.workspaces = ['packages/*'];

      expect(pkg.getFeatureFlags()).toEqual({ workspaces: ['packages/*'] });
    });

    it('returns true if a package dependency from workspace root', () => {
      expect(
        loadPackage(
          'common',
          new Project(getFixturePath('workspaces-feature-flags-root')),
        ).getFeatureFlags(),
      ).toEqual({
        decorators: false,
        flow: true,
        react: true,
        strict: true,
        typescript: true,
        workspaces: [],
      });
    });

    describe('react', () => {
      it('returns true if a package dependency (normal)', () => {
        expect(loadPackage('react').getFeatureFlags()).toEqual({ react: true, workspaces: [] });
      });
    });

    describe('typescript', () => {
      it('returns true if a package dependency (peer)', () => {
        expect(loadPackage('ts').getFeatureFlags()).toEqual({
          decorators: false,
          strict: false,
          typescript: true,
          workspaces: [],
        });
      });

      it('returns true if package contains a `tsconfig.json`', () => {
        expect(loadPackage('ts-config').getFeatureFlags()).toEqual({
          decorators: true,
          strict: true,
          typescript: true,
          workspaces: [],
        });
      });

      it('extracts decorators and strict support from local `tsconfig.json`', () => {
        expect(loadPackage('ts-config').getFeatureFlags()).toEqual({
          decorators: true,
          strict: true,
          typescript: true,
          workspaces: [],
        });
      });
    });

    describe('flow', () => {
      it('returns true if a package dependency (dev)', () => {
        expect(loadPackage('flow').getFeatureFlags()).toEqual({ flow: true, workspaces: [] });
      });
    });
  });

  describe('isComplete()', () => {
    it('returns true if every artifact is complete', () => {
      const [a, b, c] = createTestArtifacts();

      pkg.addArtifact(a);
      pkg.addArtifact(b);
      pkg.addArtifact(c);

      a.state = 'passed';
      b.state = 'passed';
      c.state = 'failed';

      expect(pkg.isComplete()).toBe(true);
    });

    it('returns false if any artifact is not complete', () => {
      const [a, b, c] = createTestArtifacts();

      pkg.addArtifact(a);
      pkg.addArtifact(b);
      pkg.addArtifact(c);

      a.state = 'passed';
      b.state = 'passed';
      c.state = 'pending';

      expect(pkg.isComplete()).toBe(false);
    });
  });

  describe('isRunning()', () => {
    it('returns true if any artifact is running', () => {
      const [a, b, c] = createTestArtifacts();

      pkg.addArtifact(a);
      pkg.addArtifact(b);
      pkg.addArtifact(c);

      a.state = 'passed';
      b.state = 'passed';
      c.state = 'building';

      expect(pkg.isRunning()).toBe(true);
    });

    it('returns false if all artifacts are not running', () => {
      const [a, b, c] = createTestArtifacts();

      pkg.addArtifact(a);
      pkg.addArtifact(b);
      pkg.addArtifact(c);

      a.state = 'passed';
      b.state = 'passed';
      c.state = 'passed';

      expect(pkg.isRunning()).toBe(false);
    });
  });

  describe('setConfigs()', () => {
    it('sets default formats for `browser` platform', () => {
      pkg.setConfigs([
        {
          format: [],
          inputs: {},
          platform: 'browser',
          namespace: '',
          support: 'stable',
        },
      ]);

      expect(pkg.configs).toEqual([
        {
          formats: ['lib', 'esm'],
          inputs: {},
          platforms: ['browser'],
          namespace: '',
          support: 'stable',
        },
      ]);
    });

    it('adds `umd` default format when `namespace` is provided for `browser` platform', () => {
      pkg.setConfigs([
        {
          format: [],
          inputs: {},
          platform: 'browser',
          namespace: 'test',
          support: 'stable',
        },
      ]);

      expect(pkg.configs).toEqual([
        {
          formats: ['lib', 'esm', 'umd'],
          inputs: {},
          platforms: ['browser'],
          namespace: 'test',
          support: 'stable',
        },
      ]);
    });

    it('sets default formats for `native` platform', () => {
      pkg.setConfigs([
        {
          format: [],
          inputs: {},
          platform: 'native',
          namespace: '',
          support: 'stable',
        },
      ]);

      expect(pkg.configs).toEqual([
        {
          formats: ['lib'],
          inputs: {},
          platforms: ['native'],
          namespace: '',
          support: 'stable',
        },
      ]);
    });

    it('sets default formats for `node` platform', () => {
      pkg.setConfigs([
        {
          format: [],
          inputs: {},
          platform: 'node',
          namespace: '',
          support: 'stable',
        },
      ]);

      expect(pkg.configs).toEqual([
        {
          formats: ['lib'],
          inputs: {},
          platforms: ['node'],
          namespace: '',
          support: 'stable',
        },
      ]);
    });

    it('doesnt set default formats when supplied manually', () => {
      pkg.setConfigs([
        {
          format: ['cjs', 'mjs'],
          inputs: {},
          platform: 'node',
          namespace: '',
          support: 'stable',
        },
      ]);

      expect(pkg.configs).toEqual([
        {
          formats: ['cjs', 'mjs'],
          inputs: {},
          platforms: ['node'],
          namespace: '',
          support: 'stable',
        },
      ]);
    });

    it('errors if invalid format is provided for `browser` platform', () => {
      expect(() => {
        pkg.setConfigs([
          {
            format: ['mjs'],
            platform: 'browser',
          },
        ]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if invalid format is provided for `native` platform', () => {
      expect(() => {
        pkg.setConfigs([
          {
            format: ['esm'],
            platform: 'native',
          },
        ]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if invalid format is provided for `node` platform', () => {
      expect(() => {
        pkg.setConfigs([
          {
            format: ['umd'],
            platform: 'node',
          },
        ]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if input name contains a slash', () => {
      expect(() => {
        pkg.setConfigs([
          {
            inputs: {
              'foo/bar': 'src/foo.ts',
            },
          },
        ]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if input name contains a space', () => {
      expect(() => {
        pkg.setConfigs([
          {
            inputs: {
              'foo bar': 'src/foo.ts',
            },
          },
        ]);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('syncPackageJson()', () => {
    it('writes to `package.json', async () => {
      const spy = jest.spyOn(fs, 'writeJson').mockImplementation();

      await pkg.syncPackageJson();

      expect(spy).toHaveBeenCalledWith(
        pkg.packageJsonPath.path(),
        { name: 'flag-common' },
        { spaces: 2 },
      );
    });
  });

  describe('tsconfigJson()', () => {
    it('returns undefined if no `tsconfig.json` in package', () => {
      expect(pkg.tsconfigJson).toBeUndefined();
    });

    it('returns options from `tsconfig.json` in package', () => {
      pkg = loadPackage('ts-config');

      expect(pkg.tsconfigJson).toEqual(
        expect.objectContaining({
          options: {
            configFilePath: pkg.path.append('tsconfig.json').path(),
            experimentalDecorators: true,
            lib: ['lib.esnext.d.ts'],
            strict: true,
          },
        }),
      );
    });
  });
});
