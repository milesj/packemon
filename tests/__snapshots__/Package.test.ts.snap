// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Package setConfigs() errors if input name contains a slash 1`] = `"Invalid flag-common field \\"inputs\\". Input \\"foo/bar\\" may only contain alpha-numeric characters."`;

exports[`Package setConfigs() errors if input name contains a space 1`] = `"Invalid flag-common field \\"inputs\\". Input \\"foo bar\\" may only contain alpha-numeric characters."`;

exports[`Package setConfigs() errors if invalid format is provided for \`browser\` platform 1`] = `
"Invalid flag-common field \\"format\\". Type must be one of: array<string>, string. Received array with the following invalidations:
 - Invalid flag-common field \\"format[0]\\". Invalid field \\"format[0]\\". String must be one of: lib, esm, umd"
`;

exports[`Package setConfigs() errors if invalid format is provided for \`native\` platform 1`] = `
"Invalid flag-common field \\"format\\". Type must be one of: array<string>, string. Received array with the following invalidations:
 - Invalid flag-common field \\"format[0]\\". Invalid field \\"format[0]\\". String must be one of: lib"
`;

exports[`Package setConfigs() errors if invalid format is provided for \`node\` platform 1`] = `
"Invalid flag-common field \\"format\\". Type must be one of: array<string>, string. Received array with the following invalidations:
 - Invalid flag-common field \\"format[0]\\". Invalid field \\"format[0]\\". String must be one of: lib, mjs, cjs"
`;
